skfda/representation/basis.py

```diff
@@ -1340,69 +1236,102 @@ def _compute_matrix(self, eval_points, derivative=0):
 
-        # normalise
-        mat = mat / np.sqrt(self.period / 2)
-        return mat
+        res = np.concatenate((constant_basis, res))
 
-    def _ndegenerated(self, penalty_degree):
-        """Return number of 0 or nearly 0 eigenvalues of the penalty matrix.
+        return res
 
-        Args:
-            penalty_degree (int): Degree of the derivative used in the
-                calculation of the penalty matrix.
+    def _penalty_orthonormal(self, weights):
+        """
+        Return the penalty when the basis is orthonormal.
+        """
 
-        Returns:
-             int: number of close to 0 eigenvalues.
+        signs = np.array([1, 1, -1, -1])
+        signs_expanded = np.tile(signs, len(weights) // 4 + 1)
 
-        """
-        return 0 if penalty_degree == 0 else 1
+        signs_odd = signs_expanded[:len(weights)]
+        signs_even = signs_expanded[1:len(weights) + 1]
+
+        phases = (np.arange(1, (self.n_basis - 1) // 2 + 1) *
+                  2 * np.pi / self.period)
+
+        # Compute increasing powers
+        coefs_no_sign = np.vander(phases, len(weights), increasing=True)
+
+        coefs_no_sign *= weights
+
+        coefs_odd = signs_odd * coefs_no_sign
+        coefs_even = signs_even * coefs_no_sign
+
+        # After applying the linear differential operator to a sinusoidal
+        # element of the basis e, the result can be expressed as
+        # A e + B e*, where e* is the other basis element in the pair
+        # with the same phase
+
+        odd_sin_coefs = np.sum(coefs_odd[:, ::2], axis=1)
+        odd_cos_coefs = np.sum(coefs_odd[:, 1::2], axis=1)
+
+        even_cos_coefs = np.sum(coefs_even[:, ::2], axis=1)
+        even_sin_coefs = np.sum(coefs_even[:, 1::2], axis=1)
+
+        # The diagonal is the inner product of A e + B e*
+        # with itself. As the basis is orthonormal, the cross products e e*
+        # are 0, and the products e e and e* e* are one.
+        # Thus, the diagonal is A^2 + B^2
+        # All elements outside the main diagonal are 0
+        main_diag_odd = odd_sin_coefs**2 + odd_cos_coefs**2
+        main_diag_even = even_sin_coefs**2 + even_cos_coefs**2
+
+        # The main diagonal should intercalate both diagonals
+        main_diag = np.array((main_diag_odd, main_diag_even)).T.ravel()
+
+        penalty_matrix = np.diag(main_diag)
+
+        # Add row and column for the constant
+        penalty_matrix = np.pad(penalty_matrix, pad_width=((1, 0), (1, 0)),
+                                mode='constant')
+
+        penalty_matrix[0, 0] = weights[0]**2
+
+        return penalty_matrix
+
+    def _penalty(self, lfd):
+
+        weights = lfd.constant_weights()
+        if weights is None:
+            return NotImplemented
+
+        # If the period and domain range are not the same, the basis functions
+        # are not orthogonal
+        if self.period != (self.domain_range[0][1] - self.domain_range[0][0]):
```

Comment:
In this case, can´t we use inner_product to compute the penalty matrix?

Reply:
When the function returns NotImplemented, it will do a numerical approach. For orthonormal Fourier basis, computing the penalty matrix when the linear differential operator has constant weights is easily done analytically, but I am not sure of how it should be done in other cases.

The numerical approach cannot use the inner product code (unless we generalize this code to accept arbitrary callables) as the inner product inside is not between things expressed in basis nor in grid form.